test/test_euler.py

```diff
@@ -0,0 +1,472 @@
+THE SOFTWARE.
+"""
+
+import numpy as np
+import numpy.linalg as la  # noqa
+import pyopencl as cl
+import pyopencl.clrandom
+import pyopencl.clmath
+from pytools.obj_array import (
+    join_fields,
+    make_obj_array,
+    with_object_array_or_scalar,
+)
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+
+# TODO: Remove grudge dependence?
+from grudge.eager import with_queue
+from grudge.symbolic.primitives import TracePair
+from mirgecom.euler import inviscid_operator
+from mirgecom.initializers import Vortex2D
+from mirgecom.initializers import Lump
+from mirgecom.boundary import BoundaryBoss
+from meshmode.discretization import Discretization
+from grudge.eager import EagerDGDiscretization
+from pyopencl.tools import (  # noqa
+    pytest_generate_tests_for_pyopencl as pytest_generate_tests,
+)
+
+# Tests go here
+
+
+def test_inviscid_flux():
+
+    ctx = cl.create_some_context(interactive=False)
+    queue = cl.CommandQueue(ctx)
+
+    def scalevec(scalar, vec):
+        # workaround for object array behavior
+        return make_obj_array([ni * scalar for ni in vec])
+
+    cl_ctx = cl.create_some_context()
+    queue = cl.CommandQueue(cl_ctx)
+
+    dim = 2
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    mesh = generate_regular_rect_mesh(
+        a=(-0.5,) * dim, b=(0.5,) * dim, n=(nel_1d,) * dim
+    )
+
+    order = 3
+
+    if dim == 2:
+        # no deep meaning here, just a fudge factor
+        dt = 0.75 / (nel_1d * order ** 2)
+    elif dim == 3:
+        # no deep meaning here, just a fudge factor
+        dt = 0.45 / (nel_1d * order ** 2)
+    else:
+        raise ValueError("don't have a stable time step guesstimate")
+
+    print("%d elements" % mesh.nelements)
+
+    discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+
+    rho = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    rhoE = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    rhoV = make_obj_array(
+        [
+            cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+            for i in range(dim)
+        ]
+    )
+    scal1 = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    p = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    ke = 0.5 * np.dot(rhoV, rhoV) / rho
+
+    # ideal single spec
+    gamma = 1.4
+    p = (gamma - 1.0) * (rhoE - ke)
+    escale = (rhoE + p) / rho
+    expected_mass_flux = rhoV
+    expected_energy_flux = scalevec(escale, rhoV)
+    expected_mom_flux = make_obj_array(
+        [
+            (rhoV[i] * rhoV[j] / rho + (p if i == j else 0))
+            for i in range(dim)
+            for j in range(dim)
+        ]
+    )
+
+    q = join_fields(rho, rhoE, rhoV)
+
+    from mirgecom.euler import _inviscid_flux
+
+    flux = _inviscid_flux(discr, q)
+
+    rhoflux = flux[0:dim]
+    rhoEflux = flux[dim : 2 * dim]
+    momflux = flux[2 * dim :]
+
+    # these should be exact, right?
```

Comment:
Why would that be the case?

Reply:
How could it not be?

Reply:
I'm not sure it's clear to me why they should necessarily be bitwise the same, if they're computed by two different bits of code.

Reply:
The two bits of code have identical operations.  TDD naturally led me to this test as I laid down and tested flux bit-by-flux bit. It (i.e. the first test in this routine) seems like a useless test now that it is all done and passing - but it does serve to illustrate succinctly exactly what we expect from the inviscid_flux routine.